File: demo/my-go-app-e2e/api_test.go
package main

import (
	"bytes"
	"context"
	"encoding/json"
	"fmt"
	"net/http"
	"testing"
	"time"

	"github.com/jackc/pgx/v5"
	"github.com/stretchr/testify/assert"
	"github.com/stretchr/testify/require"
)

var cfg Config

// TestMain runs once before all tests
func TestMain(m *testing.M) {
	cfg = LoadConfig()
	m.Run()
}

// Author represents the API response structure
type Author struct {
	ID        int       `json:"id"`
	Name      string    `json:"name"`
	Email     string    `json:"email"`
	CreatedAt time.Time `json:"created_at"`
	Books     []Book    `json:"books,omitempty"`
}

// Book represents a book in the API response
type Book struct {
	ID        int     `json:"id"`
	AuthorID  int     `json:"author_id"`
	Title     string  `json:"title"`
	Published bool    `json:"published"`
	Price     float64 `json:"price"`
}

// CreateAuthorRequest is the request payload for creating an author
type CreateAuthorRequest struct {
	Name  string              `json:"name"`
	Email string              `json:"email"`
	Books []CreateBookRequest `json:"books,omitempty"`
}

// CreateBookRequest is the request payload for creating a book
type CreateBookRequest struct {
	Title     string  `json:"title"`
	Published bool    `json:"published"`
	Price     float64 `json:"price"`
}

// =============================================================================
// GET /authors Tests
// =============================================================================

func TestGetAuthors_ReturnsOK(t *testing.T) {
	resp, err := http.Get(cfg.APIBaseURL + "/authors")
	require.NoError(t, err)
	defer resp.Body.Close()

	assert.Equal(t, http.StatusOK, resp.StatusCode)
	assert.Equal(t, "application/json", resp.Header.Get("Content-Type"))
}

func TestGetAuthors_ReturnsValidJSON(t *testing.T) {
	resp, err := http.Get(cfg.APIBaseURL + "/authors")
	require.NoError(t, err)
	defer resp.Body.Close()

	var authors []Author
	err = json.NewDecoder(resp.Body).Decode(&authors)
	require.NoError(t, err, "response should be valid JSON array of authors")

	// Should return an array (could be empty or have seeded data)
	assert.IsType(t, []Author{}, authors)
}

func TestGetAuthors_ContainsSeededData(t *testing.T) {
	resp, err := http.Get(cfg.APIBaseURL + "/authors")
	require.NoError(t, err)
	defer resp.Body.Close()

	var authors []Author
	err = json.NewDecoder(resp.Body).Decode(&authors)
	require.NoError(t, err)

	// The app seeds data on startup, so we should have at least some authors
	// Skip this assertion if running against a fresh/empty database
	if len(authors) > 0 {
		// Verify structure of first author
		assert.NotEmpty(t, authors[0].ID)
		assert.NotEmpty(t, authors[0].Name)
		assert.NotEmpty(t, authors[0].Email)
	}
}

// =============================================================================
// POST /authors Tests
// =============================================================================

func TestCreateAuthor_ReturnsCreated(t *testing.T) {
	uniqueEmail := fmt.Sprintf("e2e-test-%d@test.com", time.Now().UnixNano())
	payload := CreateAuthorRequest{
		Name:  "E2E Test Author",
		Email: uniqueEmail,
	}
	body, _ := json.Marshal(payload)

	resp, err := http.Post(
		cfg.APIBaseURL+"/authors",
		"application/json",
		bytes.NewReader(body),
	)
	require.NoError(t, err)
	defer resp.Body.Close()

	assert.Equal(t, http.StatusCreated, resp.StatusCode)

	var created Author
	err = json.NewDecoder(resp.Body).Decode(&created)
	require.NoError(t, err)

	assert.NotZero(t, created.ID, "created author should have an ID")
	assert.Equal(t, "E2E Test Author", created.Name)
	assert.Equal(t, uniqueEmail, created.Email)

	// Cleanup
	cleanupAuthor(t, created.ID)
}

func TestCreateAuthor_WritesToDatabase(t *testing.T) {
	uniqueEmail := fmt.Sprintf("e2e-db-test-%d@test.com", time.Now().UnixNano())
	payload := CreateAuthorRequest{
		Name:  "Database Write Test",
		Email: uniqueEmail,
	}
	body, _ := json.Marshal(payload)

	// Act: call the API
	resp, err := http.Post(
		cfg.APIBaseURL+"/authors",
		"application/json",
		bytes.NewReader(body),
	)
	require.NoError(t, err)
	defer resp.Body.Close()
	require.Equal(t, http.StatusCreated, resp.StatusCode)

	var created Author
	json.NewDecoder(resp.Body).Decode(&created)

	// Assert: verify data is actually in the database
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, cfg.DatabaseURL)
	require.NoError(t, err, "should connect to database")
	defer conn.Close(ctx)

	var dbID int
	var dbName, dbEmail string
	err = conn.QueryRow(ctx,
		"SELECT id, name, email FROM authors WHERE email = $1",
		uniqueEmail,
	).Scan(&dbID, &dbName, &dbEmail)

	require.NoError(t, err, "author should exist in database")
	assert.Equal(t, created.ID, dbID)
	assert.Equal(t, "Database Write Test", dbName)
	assert.Equal(t, uniqueEmail, dbEmail)

	// Cleanup
	cleanupAuthor(t, created.ID)
}

func TestCreateAuthorWithBooks_WritesAllToDatabase(t *testing.T) {
	uniqueEmail := fmt.Sprintf("e2e-books-test-%d@test.com", time.Now().UnixNano())
	payload := CreateAuthorRequest{
		Name:  "Author With Books",
		Email: uniqueEmail,
		Books: []CreateBookRequest{
			{Title: "First Book", Published: true, Price: 19.99},
			{Title: "Second Book", Published: false, Price: 29.99},
		},
	}
	body, _ := json.Marshal(payload)

	resp, err := http.Post(
		cfg.APIBaseURL+"/authors",
		"application/json",
		bytes.NewReader(body),
	)
	require.NoError(t, err)
	defer resp.Body.Close()
	require.Equal(t, http.StatusCreated, resp.StatusCode)

	var created Author
	json.NewDecoder(resp.Body).Decode(&created)

	// Verify books are in the database
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, cfg.DatabaseURL)
	require.NoError(t, err)
	defer conn.Close(ctx)

	rows, err := conn.Query(ctx,
		"SELECT title, published, price FROM books WHERE author_id = $1 ORDER BY title",
		created.ID,
	)
	require.NoError(t, err)
	defer rows.Close()

	var books []Book
	for rows.Next() {
		var b Book
		err := rows.Scan(&b.Title, &b.Published, &b.Price)
		require.NoError(t, err)
		books = append(books, b)
	}

	require.Len(t, books, 2, "should have created 2 books")
	assert.Equal(t, "First Book", books[0].Title)
	assert.Equal(t, true, books[0].Published)
	assert.Equal(t, 19.99, books[0].Price)
	assert.Equal(t, "Second Book", books[1].Title)
	assert.Equal(t, false, books[1].Published)

	// Cleanup
	cleanupAuthor(t, created.ID)
}

func TestCreateAuthor_InvalidPayload_ReturnsBadRequest(t *testing.T) {
	resp, err := http.Post(
		cfg.APIBaseURL+"/authors",
		"application/json",
		bytes.NewReader([]byte(`{invalid json`)),
	)
	require.NoError(t, err)
	defer resp.Body.Close()

	assert.Equal(t, http.StatusBadRequest, resp.StatusCode)
}

// =============================================================================
// Round-Trip Tests (Create then Get)
// =============================================================================

func TestCreateThenGet_AuthorAppearsInList(t *testing.T) {
	uniqueEmail := fmt.Sprintf("e2e-roundtrip-%d@test.com", time.Now().UnixNano())
	payload := CreateAuthorRequest{
		Name:  "Roundtrip Test Author",
		Email: uniqueEmail,
	}
	body, _ := json.Marshal(payload)

	// Create author
	resp, err := http.Post(cfg.APIBaseURL+"/authors", "application/json", bytes.NewReader(body))
	require.NoError(t, err)
	require.Equal(t, http.StatusCreated, resp.StatusCode)

	var created Author
	json.NewDecoder(resp.Body).Decode(&created)
	resp.Body.Close()

	// Fetch all authors
	resp, err = http.Get(cfg.APIBaseURL + "/authors")
	require.NoError(t, err)
	defer resp.Body.Close()

	var authors []Author
	json.NewDecoder(resp.Body).Decode(&authors)

	// Find our author
	var found *Author
	for _, a := range authors {
		if a.Email == uniqueEmail {
			found = &a
			break
		}
	}

	require.NotNil(t, found, "created author should appear in GET /authors")
	assert.Equal(t, created.ID, found.ID)
	assert.Equal(t, "Roundtrip Test Author", found.Name)

	// Cleanup
	cleanupAuthor(t, created.ID)
}

func TestCreateAuthorWithBooks_BooksAppearInGetResponse(t *testing.T) {
	uniqueEmail := fmt.Sprintf("e2e-books-roundtrip-%d@test.com", time.Now().UnixNano())
	payload := CreateAuthorRequest{
		Name:  "Author Books Roundtrip",
		Email: uniqueEmail,
		Books: []CreateBookRequest{
			{Title: "Visible Book", Published: true, Price: 15.00},
		},
	}
	body, _ := json.Marshal(payload)

	// Create
	resp, err := http.Post(cfg.APIBaseURL+"/authors", "application/json", bytes.NewReader(body))
	require.NoError(t, err)
	require.Equal(t, http.StatusCreated, resp.StatusCode)

	var created Author
	json.NewDecoder(resp.Body).Decode(&created)
	resp.Body.Close()

	// Get all authors
	resp, err = http.Get(cfg.APIBaseURL + "/authors")
	require.NoError(t, err)
	defer resp.Body.Close()

	var authors []Author
	json.NewDecoder(resp.Body).Decode(&authors)

	// Find our author and check books
	var found *Author
	for _, a := range authors {
		if a.Email == uniqueEmail {
			found = &a
			break
		}
	}

	require.NotNil(t, found)
	require.Len(t, found.Books, 1, "author should have 1 book in GET response")
	assert.Equal(t, "Visible Book", found.Books[0].Title)

	// Cleanup
	cleanupAuthor(t, created.ID)
}

// =============================================================================
// HTTP Method Tests
// =============================================================================

func TestAuthorsEndpoint_UnsupportedMethod_Returns405(t *testing.T) {
	req, _ := http.NewRequest(http.MethodDelete, cfg.APIBaseURL+"/authors", nil)
	client := &http.Client{}
	resp, err := client.Do(req)
	require.NoError(t, err)
	defer resp.Body.Close()

	assert.Equal(t, http.StatusMethodNotAllowed, resp.StatusCode)
}

// =============================================================================
// Helpers
// =============================================================================

func cleanupAuthor(t *testing.T, authorID int) {
	t.Helper()
	ctx := context.Background()
	conn, err := pgx.Connect(ctx, cfg.DatabaseURL)
	if err != nil {
		t.Logf("Warning: could not connect for cleanup: %v", err)
		return
	}
	defer conn.Close(ctx)

	// Books are deleted via CASCADE
	_, err = conn.Exec(ctx, "DELETE FROM authors WHERE id = $1", authorID)
	if err != nil {
		t.Logf("Warning: cleanup failed for author %d: %v", authorID, err)
	}
}

File: demo/my-node-app-e2e/api.test.js
import { describe, it, before, after } from 'node:test';
import assert from 'node:assert';
import { config } from './config.js';
import pg from 'pg';

const { Client } = pg;

/**
 * Helper to make HTTP requests
 */
async function request(method, path, body = null) {
    const url = `${config.apiBaseUrl}${path}`;
    const options = {
        method,
        headers: {
            'Content-Type': 'application/json'
        }
    };
    
    if (body) {
        options.body = JSON.stringify(body);
    }
    
    const response = await fetch(url, options);
    return response;
}

/**
 * Helper to cleanup test authors from database
 */
async function cleanupAuthor(authorId) {
    const client = new Client({ connectionString: config.databaseUrl });
    try {
        await client.connect();
        // Books are deleted via CASCADE
        await client.query('DELETE FROM authors WHERE id = $1', [authorId]);
    } catch (err) {
        console.warn(`Warning: cleanup failed for author ${authorId}:`, err.message);
    } finally {
        await client.end();
    }
}

/**
 * Generate unique email for test isolation
 */
function uniqueEmail(prefix = 'e2e-test') {
    return `${prefix}-${Date.now()}-${Math.random().toString(36).slice(2)}@test.com`;
}

// =============================================================================
// GET /authors Tests
// =============================================================================

describe('GET /authors', () => {
    it('should return 200 OK', async () => {
        const res = await request('GET', '/authors');
        assert.strictEqual(res.status, 200);
    });
    
    it('should return application/json content type', async () => {
        const res = await request('GET', '/authors');
        assert.strictEqual(res.headers.get('Content-Type'), 'application/json');
    });
    
    it('should return valid JSON array', async () => {
        const res = await request('GET', '/authors');
        const data = await res.json();
        assert.ok(Array.isArray(data), 'response should be an array');
    });
    
    it('should return authors with correct structure', async () => {
        const res = await request('GET', '/authors');
        const authors = await res.json();
        
        // If there are seeded authors, check structure
        if (authors.length > 0) {
            const author = authors[0];
            assert.ok('id' in author, 'author should have id');
            assert.ok('name' in author, 'author should have name');
            assert.ok('email' in author, 'author should have email');
            assert.ok('books' in author, 'author should have books');
            assert.ok(Array.isArray(author.books), 'books should be an array');
        }
    });
});

// =============================================================================
// POST /authors Tests
// =============================================================================

describe('POST /authors', () => {
    it('should return 201 Created', async () => {
        const email = uniqueEmail();
        const res = await request('POST', '/authors', {
            name: 'E2E Test Author',
            email
        });
        
        assert.strictEqual(res.status, 201);
        
        const created = await res.json();
        await cleanupAuthor(created.id);
    });
    
    it('should return created author with ID', async () => {
        const email = uniqueEmail();
        const res = await request('POST', '/authors', {
            name: 'E2E Test Author',
            email
        });
        
        const created = await res.json();
        
        assert.ok(created.id, 'created author should have an ID');
        assert.strictEqual(created.name, 'E2E Test Author');
        assert.strictEqual(created.email, email);
        
        await cleanupAuthor(created.id);
    });
    
    it('should write author to database', async () => {
        const email = uniqueEmail('e2e-db-test');
        const res = await request('POST', '/authors', {
            name: 'Database Write Test',
            email
        });
        
        const created = await res.json();
        assert.strictEqual(res.status, 201);
        
        // Verify data is actually in the database
        const client = new Client({ connectionString: config.databaseUrl });
        try {
            await client.connect();
            const result = await client.query(
                'SELECT id, name, email FROM authors WHERE email = $1',
                [email]
            );
            
            assert.strictEqual(result.rows.length, 1, 'author should exist in database');
            assert.strictEqual(result.rows[0].id, created.id);
            assert.strictEqual(result.rows[0].name, 'Database Write Test');
            assert.strictEqual(result.rows[0].email, email);
        } finally {
            await client.end();
        }
        
        await cleanupAuthor(created.id);
    });
    
    it('should create author with books and write all to database', async () => {
        const email = uniqueEmail('e2e-books-test');
        const res = await request('POST', '/authors', {
            name: 'Author With Books',
            email,
            books: [
                { title: 'First Book', published: true, price: 19.99 },
                { title: 'Second Book', published: false, price: 29.99 }
            ]
        });
        
        const created = await res.json();
        assert.strictEqual(res.status, 201);
        
        // Verify books are in the database
        const client = new Client({ connectionString: config.databaseUrl });
        try {
            await client.connect();
            const result = await client.query(
                'SELECT title, published, price FROM books WHERE author_id = $1 ORDER BY title',
                [created.id]
            );
            
            assert.strictEqual(result.rows.length, 2, 'should have created 2 books');
            assert.strictEqual(result.rows[0].title, 'First Book');
            assert.strictEqual(result.rows[0].published, true);
            assert.strictEqual(parseFloat(result.rows[0].price), 19.99);
            assert.strictEqual(result.rows[1].title, 'Second Book');
            assert.strictEqual(result.rows[1].published, false);
        } finally {
            await client.end();
        }
        
        await cleanupAuthor(created.id);
    });
    
    it('should return 400 for invalid JSON payload', async () => {
        const res = await fetch(`${config.apiBaseUrl}/authors`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: '{invalid json'
        });
        
        assert.strictEqual(res.status, 400);
    });
});

// =============================================================================
// Round-Trip Tests (Create then Get)
// =============================================================================

describe('Round-trip tests', () => {
    it('should show created author in GET /authors', async () => {
        const email = uniqueEmail('e2e-roundtrip');
        
        // Create author
        const createRes = await request('POST', '/authors', {
            name: 'Roundtrip Test Author',
            email
        });
        const created = await createRes.json();
        assert.strictEqual(createRes.status, 201);
        
        // Fetch all authors
        const getRes = await request('GET', '/authors');
        const authors = await getRes.json();
        
        // Find our author
        const found = authors.find(a => a.email === email);
        
        assert.ok(found, 'created author should appear in GET /authors');
        assert.strictEqual(found.id, created.id);
        assert.strictEqual(found.name, 'Roundtrip Test Author');
        
        await cleanupAuthor(created.id);
    });
    
    it('should show created books in GET /authors response', async () => {
        const email = uniqueEmail('e2e-books-roundtrip');
        
        // Create author with books
        const createRes = await request('POST', '/authors', {
            name: 'Author Books Roundtrip',
            email,
            books: [
                { title: 'Visible Book', published: true, price: 15.00 }
            ]
        });
        const created = await createRes.json();
        assert.strictEqual(createRes.status, 201);
        
        // Fetch all authors
        const getRes = await request('GET', '/authors');
        const authors = await getRes.json();
        
        // Find our author and check books
        const found = authors.find(a => a.email === email);
        
        assert.ok(found, 'created author should appear in GET /authors');
        assert.strictEqual(found.books.length, 1, 'author should have 1 book in GET response');
        assert.strictEqual(found.books[0].title, 'Visible Book');
        
        await cleanupAuthor(created.id);
    });
});

// =============================================================================
// HTTP Method Tests
// =============================================================================

describe('HTTP methods', () => {
    it('should return 405 for DELETE method', async () => {
        const res = await fetch(`${config.apiBaseUrl}/authors`, {
            method: 'DELETE'
        });
        
        assert.strictEqual(res.status, 405);
    });
    
    it('should return 405 for PUT method', async () => {
        const res = await fetch(`${config.apiBaseUrl}/authors`, {
            method: 'PUT'
        });
        
        assert.strictEqual(res.status, 405);
    });
});
